### PR TITLE
Log4j2: add option to fill code attributes

### PR DESCRIPTION
**Pull request by laurit**
Resolves https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/12540

### CODE REVIEW

**Comment by trask** [comment]
cc @jeanbisutti in case you have a chance to review

File: instrumentation/log4j/log4j-appender-2.17/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/log4j/appender/v2_17/Log4jHelper.java [hunk 1]
```diff
@@ -86,11 +99,53 @@ public static void capture(
...
+    // since 2.9.0
+    Class<?> stackTraceClass = null;
+    try {
+      stackTraceClass = Class.forName("org.apache.logging.log4j.util.StackLocatorUtil");
+    } catch (ClassNotFoundException exception) {
+      // ignore
+    }
+    if (stackTraceClass == null) {
+      try {
+        stackTraceClass = Class.forName("org.apache.logging.log4j.core.impl.Log4jLogEvent");
```
Comment (jeanbisutti) [inline instrumentation/log4j/log4j-appender-2.17/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/log4j/appender/v2_17/Log4jHelper.java:130]: The stack trace class is not the same following the Log4j version? If so, could you please add comments to explain this?  

It's perhaps the intent of the comment added at line 121.? If so, could you please move this comment around the line 124?

File: instrumentation/log4j/log4j-appender-2.17/library/src/main/java/io/opentelemetry/instrumentation/log4j/appender/v2_17/internal/LogEventMapper.java [hunk 2]
```diff
@@ -116,6 +125,22 @@ public void mapLogEvent(
       attributes.put(THREAD_ID, threadId);
     }
 
+    if (captureCodeAttributes) {
+      StackTraceElement source = sourceSupplier.get();
+      if (source != null) {
+        String fileName = source.getFileName();
+        if (fileName != null) {
```
Comment (jeanbisutti) [inline instrumentation/log4j/log4j-appender-2.17/library/src/main/java/io/opentelemetry/instrumentation/log4j/appender/v2_17/internal/LogEventMapper.java:132]: (optional)

```suggestion
        // file name can be null from the StackTraceElement Javadoc
        if (fileName != null) {
```

File: instrumentation/log4j/log4j-appender-2.17/library/src/test/java/io/opentelemetry/instrumentation/log4j/appender/v2_17/AbstractOpenTelemetryAppenderTest.java [hunk 3]
```diff
@@ -255,12 +262,35 @@ void logStructuredDataMessage() {
...
+  private static List<AttributeAssertion> addLocationAttributes(
+      String methodName, AttributeAssertion... assertions) {
+    return addLocationAttributes(AbstractOpenTelemetryAppenderTest.class, methodName, assertions);
+  }
+
+  protected static List<AttributeAssertion> addLocationAttributes(
+      Class<?> testClass, String methodName, AttributeAssertion... assertions) {
+    String selector = System.getProperty("Log4j2.contextSelector");
+    boolean async = selector != null && selector.endsWith("AsyncLoggerContextSelector");
+    if (async) {
```
Comment (jeanbisutti) [inline instrumentation/log4j/log4j-appender-2.17/library/src/test/java/io/opentelemetry/instrumentation/log4j/appender/v2_17/AbstractOpenTelemetryAppenderTest.java:282]: I have not found the test using an asynch appender. Could you please give me a pointer?

The Readme should perhaps explain something about the asynch appenders?

File: instrumentation/log4j/log4j-appender-2.17/library/src/main/java/io/opentelemetry/instrumentation/log4j/appender/v2_17/internal/LogEventMapper.java (hunk 2 above)
Comment (laurit) [reply instrumentation/log4j/log4j-appender-2.17/library/src/main/java/io/opentelemetry/instrumentation/log4j/appender/v2_17/internal/LogEventMapper.java:132]: This is copy-pasted from the logback instrumentation. Apparently you are the author of it.

File: instrumentation/log4j/log4j-appender-2.17/library/src/test/java/io/opentelemetry/instrumentation/log4j/appender/v2_17/AbstractOpenTelemetryAppenderTest.java (hunk 3 above)
Comment (laurit) [reply instrumentation/log4j/log4j-appender-2.17/library/src/test/java/io/opentelemetry/instrumentation/log4j/appender/v2_17/AbstractOpenTelemetryAppenderTest.java:282]: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/29da88e7cf6b8898805c0c8ddc8095477d475ce6/instrumentation/log4j/log4j-appender-2.17/javaagent/build.gradle.kts#L48-L54

Comment (jeanbisutti) [reply instrumentation/log4j/log4j-appender-2.17/library/src/test/java/io/opentelemetry/instrumentation/log4j/appender/v2_17/AbstractOpenTelemetryAppenderTest.java]: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/99f1f827a09aa5e15407746df09bbf247df605bf/instrumentation/log4j/log4j-appender-2.17/library/src/test/java/io/opentelemetry/instrumentation/log4j/appender/v2_17/AbstractOpenTelemetryAppenderTest.java#L283 replies to my second point.

**Review by jeanbisutti — APPROVED** [review]

**Pull request merged**